Perform cleanup phase even if open relation failed during drop phase of vacuum.

asubramanya opened the pull request:
Vacuum of AO tables happens in four phases: prepare, compact, drop and cleanup.
When there are concurrent transactions executing vacuum on AO tables, if one of
the vacuum transaction is executing a drop phase, the other transaction will
not be able to open the relation if it is also executing the drop phase. When
the relation open fails, vacuum simply exits without performing the cleanup
phase. The relfrozenxid in pg_class is updated during the cleanup phase and
hence the age of the table does not get reduced. This becomes a problem because
the age of the database does not reduce and eventually the xid wraparound issue
will be encountered.

This problem is fixed by performing the cleanup phase even if the open relation
fails during drop phase.

Signed-off-by: Xin Zhang <xzhang@pivotal.io>

Eulerizeit commented:
@hlinnaka @danielgustafsson could one of you review this please?

hlinnaka commented:
Hmm. Is it safe to advance relfrozenxid, if the old segment cannot be dropped? The AO segment table will still contain a row for the old segment, with an old XID, right? Or do we rely on the relfrozenxid of the AO segment table's pg_class entry for that?

I guess the bigger question here is: what does relfrozenxid mean for an AO table, anyway? There are no XIDs as such in an AO table, even though there are in the auxiliary heap tables.

ashwinstar commented:
On Thu, May 4, 2017 at 7:39 AM, Heikki Linnakangas <notifications@github.com
> wrote:
>
> I guess the bigger question here is: what does relfrozenxid mean for an AO
> table, anyway? There are no XIDs as such in an AO table, even though there
> are in the auxiliary heap tables.
>
I agree to Heikki's point, don't see reason for having relfrozenxid for AO
table itself. Maybe set it to `FrozenTransactionId` at creation time or
ignore completely from age calculations same as PT tables. Auxiliary heap
tables can maintain and reflect their own state of affairs.

ashwinstar commented on src/backend/commands/vacuum.c:
```diff
@@ -1078,9 +1082,30 @@ vacuumStatement_Relation(VacuumStmt *vacstmt, Oid relid,
+					onerel = open_relation_and_check_permission(vacstmt, relid, RELKIND_RELATION, false);
+					skip_to_cleanup = true;
+				}
+				if (!skip_to_cleanup)
```
just plain `else` is fine here, correct ?

ashwinstar commented:
On Thu, May 4, 2017 at 7:39 AM, Heikki Linnakangas <notifications@github.com
> wrote:

> Hmm. Is it safe to advance relfrozenxid, if the old segment cannot be
> dropped? The AO segment table will still contain a row for the old segment,
> with an old XID, right? Or do we rely on the relfrozenxid of the AO segment
> table's pg_class entry for that?
>
relfrozenxid is not advanced blindly, based on some predetermined value.
The change as per I understand is trying to call cleanup phase which
performs vacuum for auxiliary tables of AO and set the relfrozenxid
correctly for it even if the segment file cannot be dropped for AO table.

ashwinstar commented on src/test/regress/sql/uao_compaction/parallel_drop.sql:
```diff
@@ -0,0 +1,38 @@
+-- the AO table gets reduced correctly.
+
+-- start_ignore
+\! gpconfig -c vacuum_freeze_min_age -v 100
```
Why gpconfig and not just `SET vacuum_freeze_min_age = 100;` ?

ashwinstar replied:
Why 100? Can you comment on that. Can we have lower value ?

ashwinstar commented on src/test/regress/sql/uao_compaction/parallel_drop.sql:
```diff
@@ -0,0 +1,38 @@
+\! for i in {1..400}; do psql -q -c "insert into increase_age_table values(1)" -d regression; done;
+
+\! psql -c "vacuum ao1" regression &
+select pg_sleep(1);
```
Why ?

ashwinstar commented on src/test/regress/sql/uao_compaction/parallel_drop.sql:
```diff
@@ -0,0 +1,38 @@
+update ao1 set b = b + 1;
+
+\! gpfaultinjector -f vacuum_relation_open_relation_during_drop_phase -y reset -s 2
+\! gpfaultinjector -f vacuum_relation_open_relation_during_drop_phase -y suspend -s 2
```
I generally like to write the test in logical flow it should be, like suspending the fault injector should be done closet to really where its intended to be suspend. So in this case seems just before running vacuum.

ashwinstar commented on src/test/regress/sql/uao_compaction/parallel_drop.sql:
```diff
@@ -0,0 +1,38 @@
+update ao2 set b = b + 1;
+
+create table increase_age_table(a int);
+\! for i in {1..400}; do psql -q -c "insert into increase_age_table values(1)" -d regression; done;
```
Am I missing something, why bash way in SQL ? can't we write SQL loop here.

ashwinstar commented on src/test/regress/sql/uao_compaction/parallel_drop.sql:
```diff
@@ -0,0 +1,38 @@
+create table increase_age_table(a int);
+\! for i in {1..400}; do psql -q -c "insert into increase_age_table values(1)" -d regression; done;
+
+\! psql -c "vacuum ao1" regression &
```
This clearly is calling it needs to be in isolation framework, lets do it there.

ashwinstar commented on src/test/regress/sql/uao_compaction/parallel_drop.sql:
```diff
@@ -0,0 +1,38 @@
+
+show vacuum_freeze_min_age;
+
+select 1 is_age_increased from gp_dist_random('pg_class') where relname = 'ao2' and age(relfrozenxid) > 400 and gp_segment_id = 0;
```
`is_age_increased` seems misleading name for this statement.

asubramanya commented:
@hlinnaka It should be safe to advance the `relfrozenxid` since it is computed during the cleanup phase itself and not pre computed in a previous phase.

As @ashwinstar pointed out, the `relfrozenxid` does not make sense for AO tables since we don't store the xmin/xmax values along with the tuples. The solution of not taking into account the `relfrozenxid` during the calculation of the database age will be addressed separately.

The main intent of this patch is to make sure that the auxiliary tables get vacuumed correctly and their age gets reduced regardless of whether vacuum was able to drop the segment files since they are not dependent on each other.

asimrp commented on src/backend/commands/vacuum.c:
```diff
@@ -1078,9 +1082,30 @@ vacuumStatement_Relation(VacuumStmt *vacstmt, Oid relid,
 			{
 				DeregisterSegnoForCompactionDrop(relid,
 									vacstmt->appendonly_compaction_segno);
-				CommitTransactionCommand();
```
At the end of compaction, the compacted segfile is in AWATING_DROP state in aoseg table.  For this change to be visible, `CommitTransactionCommand()` is required.  I don't see why it is made optional.

asimrp commented on src/backend/commands/vacuum.c:
```diff
@@ -1119,65 +1144,70 @@ vacuumStatement_Relation(VacuumStmt *vacstmt, Oid relid,
+						MemoryContextSwitchTo(oldctx);
+						relation_close(onerel, NoLock);
+						CommitTransactionCommand();
+						/* don't dispatch this iteration */
```
If compaction happens and drop is skipped due to serializable backends, cleanup phase needs to be executed, right?

asubramanya commented:
We rewrote the test to use isolation2 framework and also refactored the code a bit to make it easier to read and understand.
@asimrp @xinzweb Can you please take a look.

xinzweb commented on src/test/isolation2/output/uao/vacuum_cleanup.source:
```diff
@@ -0,0 +1,56 @@
+
+1: set vacuum_freeze_min_age = 0;
+SET
+1: vacuum ao_@orientation@_vacuum_cleanup2;
```
Do we have the age > 3 before the vacuum? This will indicate the age got reduced.

asubramanya replied:
Yes. We have added a `SELECT` statement to display the age to confirm that the age has increased beyond 3.

xinzweb commented on src/backend/commands/vacuum.c:
```diff
@@ -956,7 +956,6 @@ vacuumStatement_Relation(VacuumStmt *vacstmt, Oid relid,
 	Relation			onerel;
 	LockRelId			onerelid;
 	MemoryContext		oldctx;
-	bool				bTemp;
```
Glad we get rid of bTemp. :)

xinzweb commented on src/backend/commands/vacuum.c:
```diff
@@ -1072,19 +1083,40 @@ vacuumStatement_Relation(VacuumStmt *vacstmt, Oid relid,
+				 * cleanup phase so that the relfrozenxid value is updated
+				 * correctly in pg_class
+				 */
+				vacstmt->appendonly_compaction_vacuum_cleanup = true;
```
It's awesome to reuse the `_cleanup` flag here without introduce another `skip_to_cleanup` flag. :)

asimrp commented on src/test/isolation2/sql/setup.sql:
```diff
@@ -85,3 +85,20 @@ SELECT coalesce(
+            return True
+        time.sleep(0.5)
+    return False 
+$$ LANGUAGE plpythonu;
```
Nice!

asubramanya commented on src/test/isolation2/input/uao/vacuum_cleanup.source:
```diff
@@ -0,0 +1,62 @@
+-- @Description Test that when there is a parallel vacuum going on in drop phase, the age of
+-- the AO/AOCS table gets reduced correctly.
+
+1: create table ao_@orientation@_vacuum_cleanup1(a int, b int) with(appendonly=true, orientation=@orientation@);
```
The gucs will not get inherited by the sessions, so this won't work.

ashwinstar replied:
You will have to do it on per session basis for it to work, I think its not worth it for this test.

asubramanya commented:
Thanks for the review. Pushed to master.
